SubEthaEdit-Mac/Source/TabbedDocument.m

```diff
@@ -0,0 +1,145 @@
+    NSAlert *alert = [[NSAlert alloc] init];
+
+    [alert setAlertStyle:style];
+    [alert setMessageText:message];
+    [alert setInformativeText:details];
+
+    for (NSString * button in buttons) {
+        [alert addButtonWithTitle: button];
+    }
+
+    __unsafe_unretained NSDocument *weakSelf = self;
+    [self presentAlert:alert completionHandler:^(NSModalResponse returnCode) {
+        NSDocument *strongSelf = weakSelf;
+        if (strongSelf && then) {
+            then(strongSelf, returnCode);
+        }
+    }];
+}
+
+- (void)inform:(NSString *)message details:(NSString *)details {
+    [self alert:message
+          style:NSAlertStyleInformational
+        details:details
+        buttons:@[NSLocalizedString(@"OK", nil)]
+           then:nil];
+}
+
+- (void)warn:(NSString *)message
+     details:(NSString *)details
+     buttons:(NSArray *)buttons
+        then:(AlertConsequence)then {
+    [self alert:message
+          style:NSAlertStyleWarning
+        details:details
+        buttons:buttons
+           then:then];
+}
+
+- (void)presentAlert:(NSAlert *)alert completionHandler:(void (^)(NSModalResponse returnCode))completionHandler {
+    NSArray *orderedWindows = NSApp.orderedWindows;
+    NSSet *candidateWindows = [NSSet setWithArray:[self.windowControllers valueForKey:@"window"]];
+
+    NSUInteger index = [orderedWindows indexOfObjectPassingTest:
+                        ^(NSWindow *window, NSUInteger idx, BOOL *_stop) {
+                            return [candidateWindows containsObject:window];
+                        }];
+    NSWindow *window = orderedWindows[index];
+
+    [window makeKeyAndOrderFront:self];
+    [alert beginSheetModalForWindow:window completionHandler:completionHandler];
+}
+
+@end
+
+
+@implementation TabbedDocument (SheetSynchronization)
+
+// We collect all the notification names and selectors we care about in this method
+// to avoid potential bugs of forgetting to remove an observer we've added.
+- (NSDictionary *)windowNotifications {
+    return @{
+             NSWindowWillBeginSheetNotification:
+                 NSStringFromSelector(@selector(windowWillBeginSheet:)),
+             NSWindowDidEndSheetNotification:
+                 NSStringFromSelector(@selector(windowDidEndSheet:))
+             };
+}
+
+- (void)addWindowController:(NSWindowController *)windowController {
+    [super addWindowController:windowController];
+
+    NSNotificationCenter *defaultCenter = NSNotificationCenter.defaultCenter;
+    NSDictionary *notifications = self.windowNotifications;
+    for (NSString *name in notifications)
+        [defaultCenter addObserver:self
+                          selector:NSSelectorFromString(notifications[name])
+                              name:name
+                            object:windowController.window];
+}
+
+- (void)removeWindowController:(NSWindowController *)windowController {
+    [super addWindowController:windowController];
```

Comment:
not sure if the contract is that `removeWindowController` is called before dealloc. Otherwise we also need the removeal in dealloc.

Reply:
I'll look into this a bit more, but I think the only way you eventually get a `dealloc` in `NSDocument` is by closing all its windows (I guess I could construct some sort of test where you have Document's switch windows or something, but then you'd still probably be triggering `removeWindowController:`). Either way, SEE's document subclass seems to indiscriminately `removeObserver:self` in it's dealloc, so it may be overdetermined.